src/main/utility/synchronization/count_down_latch.rs

```diff
@@ -0,0 +1,376 @@
+use std::sync::{Arc, Condvar, Mutex};
+
+/// A latch counter.
+///
+/// If a counter is cloned, it will inherit the counter's state for the current generation. For
+/// example if a counter is cloned after it has already counted down, then the new counter will also
+/// be treated as if it had already counted down in the current generation. If a counter is cloned
+/// before it has counted down, then the new counter will also need to count down in the current
+/// generation.
+#[derive(Debug)]
+pub struct LatchCounter {
+    inner: Arc<LatchInner>,
+    /// This counter's generation.
```

Comment:
Changed the wording. It's more like an ID than the number of rounds since it will eventually wrap (and for example using less-than or greater-than would lead to errors), so I used "an ID for this counter's count-down round" instead.